src/misc.jl

```diff
@@ -68,10 +68,10 @@ function concrete_length(xs)
     n
 end
 
-function concrete_length(xs::DataArray{T}) where T
+function concrete_length(xs::Array{Union{Missing,T}}) where T
```

Comment:
To cover all cases where an array might contain missing values, this needs to be `AbstractArray{>:Missing}`. Even better, you can just remove this method and use `!ismissing(x) && isconcrete(x)` in the generic method above; the  compiler will optimize the `ismissing` call out when the input does not allow for missing values.

The same remark applies to other functions below. Basically, `::Array{Union{Missing,T}}` should never appear in signatures as it's too restrictive. Overall that should also allow dropping a lot of duplicated code.